### ISSUE TITLE
Runtime error/typechecker: class constraint not passed along

### ISSUE THREAD

**Issue by naehrstoff** [purescript/purescript#2443]
I'm quite new to Purescript (I have a JS background), so I did not yet write a test case for this. If it turns out that this is indeed an issue that has not been resolved yet (I looked through the open type checker bugs but didn't find something similar), then I might be able to provide a test case with a bit of help.

I base my example on `purescript-css` because I fear that I would introduce subtle errors if I invent my own example and I think this is not a `purescript-css` bug. That's why I inline some type definitions from `purescript-css` for reference:

```purescript
-- Full type definitions from purescript-css for reference
newtype StyleM a = S (Writer (Array Rule) a)
newtype Value = Value …
newtype JustifyContentValue = JustifyContentValue Value
class Center   a where center   :: a
```

The actual problem is this (I basically wanted to create an alias for an existing function):

```purescript
justifyContent :: JustifyContentValue -> StyleM

justifyContentCenter :: forall a. (Center a) => a
justifyContentCenter = center

myStyle :: StyleM Unit
myStyle = do
  justifyContent justifyContentCenter
```

What happens is that the class constraint for `a` seems to not be passed along, resulting in a runtime error:

```javascript
// Runtime error "Uncaught Error: Failed pattern match at CSS.Render …"
// In the call below v.value1 is not defined, but should be
CSS_Render.collect(new Data_Tuple.Tuple(v.value0, v.value1)))
```

When I provide a concrete type it works:

```purescript
justifyContentCenter :: JustifyContentValue
justifyContentCenter = center
```

**Comment by LiamGoodacre** [comment]
Are you directly calling `CSS.Render.collect`?  As I can't see in `purescript-css` any place where the code `CSS_Render.collect(...)` would be generated.  Are you using any other libraries and at which versions?

**Comment by naehrstoff** [comment]
Yes, sorry, some version info:

* `psc` 0.10.2
* `purescript-pux` 7.0.0
* `purescript-pux-css` 3.1.0
* `purescript-dom` 3.1.0

The call to `CSS.Render.collect` is where it blows up in the end, but I don't think it has anything to do with the problem because of the missing `v.value1`. Or do you think?

There's also this bit of glue code that we use to translate from `purescript-css` to a format that [Aphrodite](https://github.com/Khan/aphrodite) understands (we're using inline styles throughout the app):

```purescript
aphroditeCSS :: CSS -> AphroditeCSS
aphroditeCSS rules = { decls: decls, nested: nested }
  where
    decls = map tuple $ foldr tuples [] $ runS rules
    tuple p = either (\_ -> mempty) (\(Tuple k v) -> Tuple (camelCase k) v) p
    tuples (Property key val) array = collect (Tuple key val) <> array
    tuples _ array = array

    nested = reverse $ foldl nestedTuples [] $ runS rules

    nestedTuples array (Nested app nestedRules) = [ Tuple key value ] <> array
      where
        key = (selector $ merger $ NonEmpty.singleton app)
        value = (map tuple $ foldr tuples [] nestedRules)
    nestedTuples array  _                       = array
```

**Comment by LiamGoodacre** [comment]
I haven't been able to reproduce, this is what I've tried:

```purescript
module Main where

import Prelude
import Data.Tuple (Tuple(..))
import Data.Foldable (fold, foldMap)
import Control.Monad.Eff (Eff)
import Control.Monad.Eff.Console (CONSOLE, logShow)

import CSS (CSS, Rule(..), runS, collect, StyleM, justifyContent)
import CSS.Common (class Center, center)

justifyContentCenter :: forall a. (Center a) => a
justifyContentCenter = center

myStyle :: StyleM Unit
myStyle = justifyContent justifyContentCenter

go :: CSS -> Array (Tuple String String)
go = map fold <<< foldMap tuples <<< runS where
  tuples (Property key val) = collect (Tuple key val)
  tuples _ = []

main :: forall e. Eff (console :: CONSOLE | e) Unit
main = logShow (go myStyle)
```

**Comment by paf31** [comment]
I also can't reproduce this, with or without the type annotation.

**Comment by naehrstoff** [comment]
Alright, thanks for double-checking! I was able to trigger this bug consistently in our application, but there are many more moving parts there, so it might be something different.

I will try to take the code from @LiamGoodacre as a starting point and see where something special might happen. If I can't find a test case that works outside our project I will close this issue by the end of next week.

In a way that would be good: no bug! On the other hand I was still very surprised by this run-time error. Oh well 😄

**Comment by paf31** [comment]
Are you using the FFI at all?

**Comment by naehrstoff** [comment]
Yes, the FFI is indeed involved. I added the code in use below, but I think it's getting a bit unwieldy to understand what's happening, sorry about that. I will try to come up with a test case that gets to the essence of this, but it might take me a while.

```purescript
foreign import data Style :: *

aphroditeStyle :: CSS -> Style
aphroditeStyle rules = js_aphroditeStyle "style" $ aphroditeCSS rules

myStyle = aphroditeStyle do
  justifyContent justifyContentCenter
```

where

```javascript
exports.js_aphroditeStyle = function(tag) {
  return function(acss) {
    var style = {};
    acss.decls.forEach(function(r) {
      style[r.value0] = r.value1;
    });

    acss.nested.forEach(function(n) {
      var nested = {};
      n.value1.forEach(function(r) {
        nested[r.value0] = r.value1;
      });
      style[n.value0] = nested;
    });

    return exports.createStyle(tag)(style);
  };
};
```

**Comment by hdgarrood** [comment]
Can you also include the `foreign import` declaration for `js_aphroditeStyle` please?

**Comment by naehrstoff** [comment]
Yes, of course, I edited my last comment to include the full FFI code.

**Comment by hdgarrood** [comment]
It might not be the cause of this error, but if you can, I would recommend not relying on the `value0, `value1` etc properties in your FFI code, as I don't think the representation of types defined using `data` is considered part of the public API. Also, for debugging this kind of thing, I would recommend adding calls to functions like `traceAny` in various places to see where the runtime representation of things differs from what is expected.

**Comment by naehrstoff** [comment]
So I tried to extract this problem into a small application, but was not able to reproduce it. However, I was still able to reproduce this in our main application, so I investigated further.

**tl;dr** I conclude that the reason for the problem lies in the way I compile the project using `pscid` and `webpack`. I do not fully recompile the project using `pulp -w build`, for example, which leads to changes in the generated code that are not updated in the places it is used.

Here's how I can reproduce the bug:

1. Delete `/output`
2. Build with `psc 'bower_components/purescript-*/src/**/*.purs' 'src/**/*.purs'`
3. Start Webpack server
4. Open the website: no errors in the console so far
5. Start `pscid` in parallel
6. In the source file
   a) Change `justifyContentCenter :: JustifyContentValue`
   b) to `justifyContentCenter :: forall a. (Center a) => a`
7. `pscid` will recompile the file, Webpack will reload the page, the error occurs

The code looks like this in step 4, which makes sense:

```javascript
// justifyContentCenter :: JustifyContentValue
var justifyContentCenter = CSS_Common.center(CSS_Flexbox.centerJustifyContentValue);
```

When I make the changes in 6, step 7 will generate the output below. Note the type signature in the comment: it does not correctly reflect what's being generated. (I'm not sure whether the wrong comment is a problem in `pscid` or Webpack).

```javascript
/**
 * justifyContentCenter :: JustifyContentValue
 */
var justifyContentCenter = function (dictCenter) {
    return CSS_Common.center(dictCenter);
};
```

The code that is being generated here is correct, though, if we consider the type signature of `justifyContentCenter :: forall a. (Center a) => a` that had been used to generate it. But I think because I only used `pscid`, the code was only regenerated in this file, but not the other files where it is being used. This led to a function being passed in as an argument instead of a string.

When I fully compile the project again using `psc` or `pulp -w build`, everything works with either type signature.

So it seems that this was just me being ignorant of how all the tools work together. I was aware that `pscid` only recompiles one file at a time, but did not think of the effect that this had with regards to higher-kinded types (fingers crossed that this is the correct term 😄).

Thank you all for helping out and sorry that this was just a user error 😭. I think it's safe to close this issue, feel free to reopen if you need more info from me or if this should be investigated further.

**Comment by hdgarrood** [comment]
Should this be considered a bug in `pscid`?

**Comment by paf31** [comment]
I don't think so, `pscid` is only meant to recompile one file at a time by default.

**Comment by wereHamster** [comment]
FWIW, ghcid recompiles all files in the project which were affected by the change. So you get the full picture whether the project typechecks as a whole or not.
